jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,517 @@
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.io.CyclicTimeouts;
+import org.eclipse.jetty.server.ConnectionMetaData;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+import org.eclipse.jetty.util.annotation.Name;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.eclipse.jetty.util.thread.Scheduler;
+
+/**
+ * A Denial of Service Handler.
+ * <p>Protect from denial of service attacks by limiting the request rate from remote hosts</p>
+ */
+@ManagedObject("DOS Prevention Handler")
+public class DosHandler extends ConditionalHandler.ElseNext
+{
+    /**
+     * An id {@link Function} to create an ID from the remote address and port of a {@link Request}
+     */
+    public static final Function<Request, String> ID_FROM_REMOTE_ADDRESS_PORT = request ->
+    {
+        SocketAddress remoteSocketAddress = request.getConnectionMetaData().getRemoteSocketAddress();
+        if (remoteSocketAddress instanceof InetSocketAddress inetSocketAddress)
+            return inetSocketAddress.toString();
+        return remoteSocketAddress.toString();
+    };
+
+    /**
+     * An id {@link Function} to create an ID from the remote address of a {@link Request}
+     */
+    public static final Function<Request, String> ID_FROM_REMOTE_ADDRESS = request ->
+    {
+        SocketAddress remoteSocketAddress = request.getConnectionMetaData().getRemoteSocketAddress();
+        if (remoteSocketAddress instanceof InetSocketAddress inetSocketAddress)
+            return inetSocketAddress.getAddress().toString();
+        return remoteSocketAddress.toString();
+    };
+
+    /**
+     * An id {@link Function} to create an ID from the remote port of a {@link Request}.
+     * This can be useful if there is an untrusted intermediary, where the remote port can be a surrogate for the connection.
+     */
+    public static final Function<Request, String> ID_FROM_REMOTE_PORT = request ->
+    {
+        SocketAddress remoteSocketAddress = request.getConnectionMetaData().getRemoteSocketAddress();
+        if (remoteSocketAddress instanceof InetSocketAddress inetSocketAddress)
+            return Integer.toString(inetSocketAddress.getPort());
+        return remoteSocketAddress.toString();
+    };
+
+    /**
+     * An id {@link Function} to create an ID from {@link ConnectionMetaData#getId()} of the {@link Request}
+     */
+    public static final Function<Request, String> ID_CONNECTION = request -> request.getConnectionMetaData().getId();
+
+    /**
+     * An interface implemented to track and control the rate of requests for a specific ID.
+     */
+    public interface RateControl
+    {
+        boolean isRateExceeded(long now, boolean addSample);
```

Comment:
These two methods should be javadoc'ed, as I'm not entirely clear on what exactly changes when `addSample` is true vs false.